### PR TITLE
Independent Samples Analyses Updated

### PR DESCRIPTION
**Pull request by runjin326**
### Purpose/implementation Section
1) Update `tumor_descriptor` + `experimental_strategy` to match new `histologies.tsv` release
2) Now we consider `cohort` and `cancer_group` when calling samples "independent" - e.g., for patients that are present in different cohort, they are considered two independent specimens 

#### What was your approach?
1) Related lines of codes are modified 
2) Parse data based on `cohort` and `cancer_group`, call independent samples for each `cohort` and `cancer_group` and then combine

#### What GitHub issue does your pull request address?
https://github.com/PediatricOpenTargets/ticket-tracker/issues/89

#### Which areas should receive a particularly close look?
Please double check the modified code and the differences in results and make sure they all make sense.

#### Is there anything that you want to discuss further?
For Rscript 01, if I run it line by line, it is fine but when I run it in terminal, kept getting following error: `Error in filter(., cohort == cohort_name) : object 'cohort' not found` - very curious as to why...

#### Is the analysis in a mature enough form that the resulting figure(s) and/or table(s) are ready for review?
Yes

### Results
`wc -l *.tsv
   11614 independent-specimens.rnaseq.primary-plus.tsv
   12080 independent-specimens.rnaseq.primary.tsv
     374 independent-specimens.rnaseq.relapse.tsv
    1109 independent-specimens.wgs.primary-plus.tsv
     980 independent-specimens.wgs.primary.tsv
     169 independent-specimens.wgs.relapse.tsv
    2159 independent-specimens.wgswxspanel.primary-plus.tsv
    2007 independent-specimens.wgswxspanel.primary.tsv
     207 independent-specimens.wgswxspanel.relapse.tsv
   30699 total
`
#### What types of results are included (e.g., table, figure)?
*.tsv files with results 

#### What is your summary of the results?
More specimens are now included as "independent" samples.

#### Reproducibility Checklist

- [x] The dependencies required to run the code in this pull request have been added to the project Dockerfile.
- [x] This analysis has been added to continuous integration.

#### Documentation Checklist

- [x] This analysis module has a `README` and it is up to date.
- [ ] This analysis is recorded in the table in `analyses/README.md` and the entry is up to date.
- [x] The analytical code is documented and contains comments.

### CODE REVIEW

**Comment by logstar** [comment]
> For Rscript 01, if I run it line by line, it is fine but when I run it in terminal, kept getting following error: `Error in filter(., cohort == cohort_name) : object 'cohort' not found` - very curious as to why...

**Comment by logstar** [comment]
Thank you for preparing the update @runjin326 !

To get back to your question

> For Rscript 01, if I run it line by line, it is fine but when I run it in terminal, kept getting following error: `Error in filter(., cohort == cohort_name) : object 'cohort' not found` - very curious as to why...

I also encountered this a while back, because I forgot that `dplyr` is not imported, and `stats::filter` is used by default.

File: analyses/independent-samples/run-independent-samples.sh [hunk 1]
```diff
@@ -14,7 +14,7 @@ RUN_FOR_SUBTYPING=${OPENPBTA_BASE_SUBTYPING:-0}
 # Set the working directory to the directory of this file
 cd "$(dirname "${BASH_SOURCE[0]}")"
 
-Rscript -e "rmarkdown::render('00-repeated-samples.Rmd',params=list(base_run = ${RUN_FOR_SUBTYPING}), clean = TRUE)"
+# Rscript -e "rmarkdown::render('00-repeated-samples.Rmd',params=list(base_run = ${RUN_FOR_SUBTYPING}), clean = TRUE)"
```
Comment (logstar) [inline analyses/independent-samples/run-independent-samples.sh]: If we comment this line out, should we also remove `00-repeated-samples.nb.html`, so we do not assume `00-repeated-samples.nb.html` is up-to-date?

File: analyses/independent-samples/independent-samples.R [hunk 2]
```diff
@@ -39,34 +39,43 @@ independent_samples <- function(sample_df,
...
+  cancer_group_list <- sample_df$cancer_group %>% unique()    
+  
+  independent_all <- data.frame(Kids_First_Participant_ID = character(), cohort = character(), cancer_group = character(), Kids_First_Biospecimen_ID = character(), stringsAsFactors = FALSE)
+  
+  for (i in 1:length(cohort_list)){
+    for (j in 1:length(cancer_group_list)){
+      # filter to the specific cancer group and cohort
+      cohort_name <- cohort_list[i]
+      cancer_group_name <- cancer_group_list[j]
+      filtered_df <- sample_df %>% filter(cohort == cohort_name) %>%
```
Comment (logstar) [inline analyses/independent-samples/independent-samples.R]:
```suggestion
      filtered_df <- sample_df %>% dplyr::filter(cohort == cohort_name) %>%
```

File: analyses/independent-samples/independent-samples.R [hunk 3]
```diff
@@ -39,34 +39,43 @@ independent_samples <- function(sample_df,
...
+        filter(cancer_group == cancer_group_name)
```
Comment (logstar) [inline analyses/independent-samples/independent-samples.R]:
```suggestion
        dplyr::filter(cancer_group == cancer_group_name)
```

File: analyses/independent-samples/independent_rna_samples.R [hunk 4]
```diff
@@ -126,20 +126,31 @@ independent_rna_samples <- function(independent_dna_sample_df,
...
   
-  return(early_ind)
-}
+  for (i in 1:length(cohort_list)){
+    for (j in 1:length(cancer_group_list)){
+      # filter to the specific cancer group and cohort
+      cohort_name <- cohort_list[i]
+      cancer_group_name <- cancer_group_list[j]
+      filtered_df <- sample_df %>% filter(cohort == cohort_name) %>%
+        filter(cancer_group == cancer_group_name)
```
Comment (logstar) [inline analyses/independent-samples/independent_rna_samples.R]:
```suggestion
      filtered_df <- sample_df %>% dplyr::filter(cohort == cohort_name) %>%
        dplyr::filter(cancer_group == cancer_group_name)
```

**Review by logstar — CHANGES_REQUESTED** [review]
Thank you again for updating this module @runjin326 !

The new procedures look good to me.

I was wondering if you could also generate the independent sample lists using `all-cohorts` like before for backward compatibility. I would suggest adding a `independence_level` parameter to `independent_rna_samples()` and `independent_samples()`, which takes the value of either `'all-cohorts'` or ``each-cohort``.

Sorry that I should have thought about having two types of independent lists when writing the ticket.

I was also wondering if you could use the same table format as before for all independent sample list TSVs for backward compatibility, so that each TSV would have two columns, `Kids_First_Participant_ID` and `Kids_First_Biospecimen_ID`.

If you have any questions, comments, or suggestions, feel free to let me know.

**Comment by runjin326** [comment]
Thanks Yuanchao @logstar for reviewing this. I will make the changes accordingly!

File: analyses/independent-samples/run-independent-samples.sh (hunk 1 above)
Comment (runjin326) [reply analyses/independent-samples/run-independent-samples.sh]: Sorry I commented it while I was running the script but I forgot to commented it back.

**Comment by runjin326** [comment]
@logstar Hey Yuanchao, I have made changes accordingly - please let me know whether this is what you had in mind.

File: analyses/independent-samples/01-generate-independent-specimens.R [hunk 5]
```diff
@@ -54,18 +54,31 @@ if (!dir.exists(out_dir)){
...
+wgs_relapse_all_file <- file.path(out_dir, 
+                                   "independent-specimens.wgs.relapse.allcohorts.tsv")
+wgs_primplus_all_file <- file.path(out_dir, 
+                                    "independent-specimens.wgs.primary-plus.allcohorts.tsv")
+wgswxspanel_primary_all_file <- file.path(out_dir, 
+                                           "independent-specimens.wgswxspanel.primary.allcohorts.tsv")
+wgswxspanel_relapse_all_file <- file.path(out_dir, 
+                                           "independent-specimens.wgswxspanel.relapse.allcohorts.tsv")
+wgswxspanel_primplus_all_file <- file.path(out_dir, 
+                                            "independent-specimens.wgswxspanel.primary-plus.allcohorts.tsv")
```
Comment (logstar) [inline analyses/independent-samples/01-generate-independent-specimens.R]: For backward compatibility, could you change the all-cohorts filenames to be the same as previous results?

```suggestion
wgs_primary_all_file <- file.path(out_dir, 
                                   "independent-specimens.wgs.primary.tsv")
wgs_relapse_all_file <- file.path(out_dir, 
                                   "independent-specimens.wgs.relapse.tsv")
wgs_primplus_all_file <- file.path(out_dir, 
                                    "independent-specimens.wgs.primary-plus.tsv")
wgswxspanel_primary_all_file <- file.path(out_dir, 
                                           "independent-specimens.wgswxspanel.primary.tsv")
wgswxspanel_relapse_all_file <- file.path(out_dir, 
                                           "independent-specimens.wgswxspanel.relapse.tsv")
wgswxspanel_primplus_all_file <- file.path(out_dir, 
                                            "independent-specimens.wgswxspanel.primary-plus.tsv")
```

File: analyses/independent-samples/02-generate-independent-rnaseq.R [hunk 6]
```diff
@@ -54,12 +60,19 @@ if (!dir.exists(out_dir)){
...
+                                  "independent-specimens.rnaseq.relapse.eachcohort.tsv")
+rnaseq_primplus_each_file <- file.path(out_dir, 
+                                  "independent-specimens.rnaseq.primary-plus.eachcohort.tsv")
+
+rnaseq_primary_all_file <- file.path(out_dir, 
+                                 "independent-specimens.rnaseq.primary.allcohorts.tsv")
+rnaseq_relapse_all_file <- file.path(out_dir, 
+                                 "independent-specimens.rnaseq.relapse.allcohorts.tsv")
+rnaseq_primplus_all_file <- file.path(out_dir, 
+                                  "independent-specimens.rnaseq.primary-plus.allcohorts.tsv")
```
Comment (logstar) [inline analyses/independent-samples/02-generate-independent-rnaseq.R]: Similarly, for backward compatibility.

```suggestion
rnaseq_primary_all_file <- file.path(out_dir, 
                                 "independent-specimens.rnaseq.primary.tsv")
rnaseq_relapse_all_file <- file.path(out_dir, 
                                 "independent-specimens.rnaseq.relapse.tsv")
rnaseq_primplus_all_file <- file.path(out_dir, 
                                  "independent-specimens.rnaseq.primary-plus.tsv")
```

File: analyses/independent-samples/independent-samples.R [hunk 7]
```diff
@@ -51,31 +57,50 @@ independent_samples <- function(histology_df,
...
+        # filter to the specific cancer group and cohort
+        cohort_name <- cohort_list[i]
+        cancer_group_name <- cancer_group_list[j]
+        # deal with cancer group is NA to avoid missing samples
+        if(is.na(cancer_group_name)){
+          filtered_df <- sample_df %>% dplyr::filter(cohort == cohort_name) %>%
+            dplyr::filter(is.na(cancer_group))
+        }else{
+          filtered_df <- sample_df %>% dplyr::filter(cohort == cohort_name) %>%
+            dplyr::filter(cancer_group == cancer_group_name)}
```
Comment (logstar) [inline analyses/independent-samples/independent-samples.R]: Line break before the closing `}` for readability.

**Review by logstar — CHANGES_REQUESTED** [review]
Thank you for the updates! The new procedures looks good to me.

Still for backward compatibility, could you change the all-cohorts filenames to be the same as previous results? Specific code suggestions are as following.

**Comment by runjin326** [comment]
@logstar, changes made now! :)

**Review by logstar — APPROVED** [review]
Thank you for the quick updates @runjin326 ! The PR looks good to me.

Although the [README.md](https://github.com/PediatricOpenTargets/OpenPedCan-analysis/blob/independent-samples-update/analyses/independent-samples/README.md) has not been updated for the each-cohort procedure, we could merge this PR now, so other modules can use the updated independent sample lists, and update the README.md later.

**Pull request merged**